packages/@pollyjs/adapter-node-http/src/-private/transport-wrapper.js

```diff
@@ -0,0 +1,290 @@
+      request.end();
+    });
+
+    const responseBody = await new Promise(resolve => {
+      const chunks = [];
+
+      response.on('data', chunk => chunks.push(chunk));
+      response.on('end', () =>
+        resolve(this.getBodyFromChunks(chunks, response.headers))
+      );
+    });
+
+    return {
+      headers: response.headers,
+      statusCode: response.statusCode,
+      body: responseBody
+    };
+  }
+
+  async respond(pollyRequest) {
+    const { response } = pollyRequest;
+    const [, req] = pollyRequest.requestArguments;
+    const fakeSocket = { readable: false };
+    const res = new http.IncomingMessage(fakeSocket);
+
+    res.statusCode = response.statusCode;
+    res.headers = { ...response.headers };
+    res.rawHeaders = keys(res.headers).forEach(key =>
+      res.rawHeaders.push(key, res.headers[key])
+    );
+
+    await new Promise(resolve => process.nextTick(resolve));
+
+    req.emit('response', res);
+
+    const chunks = this.getChunksFromBody(response.body, response.headers);
+
+    setImmediate(function emitChunk() {
+      const chunk = chunks.shift();
+
+      if (chunk) {
+        res.push(chunk);
+        setImmediate(emitChunk);
+      } else {
+        res.push(null);
+      }
+    });
+
+    req.emit('prefinish');
+    req.emit('finish');
+    req.emit('end');
+  }
+
+  createRequestWrapper() {
+    const wrapper = this;
+    const { adapter, transport } = wrapper;
+    const { request: nativeRequest } = nativeRequestMapping.get(transport);
+
+    return function request(...args) {
+      const req = nativeRequest.call(transport, ...args);
+      const nativeWrite = req.write;
+      const chunks = [];
+      let ended = false;
+
+      // Override req.write so we can save all the request body chunks
+      req.write = (chunk, encoding, callback) => {
+        if (!req.aborted) {
+          if (chunk) {
+            chunks.push(
+              Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk, encoding)
+            );
+          }
+        }
+
+        return nativeWrite.call(req, chunk, encoding, callback);
+      };
+
+      // Override req.end since this actual request is just a shell that never gets sent out.
+      // When we need to get actual data, a new request is made and a response
+      // is faked to this req.
+      req.end = (chunk, encoding, callback) => {
+        if (req.aborted || ended) {
+          return;
+        }
+
+        ended = true;
+
+        if (typeof chunk === 'function') {
+          callback = chunk;
+          chunk = undefined;
+        } else if (typeof encoding === 'function') {
+          callback = encoding;
+          encoding = undefined;
+        }
+
+        if (chunk) {
+          req.write(chunk, encoding);
+        }
+
+        // No need to carry callback around. This is what happens in original `end`.
+        if (typeof callback === 'function') {
+          req.once('finish', callback);
+        }
+
+        const headers =
```

Comment:
@yasinuslu I've removed the `response` listener removal here since this request never actually gets sent out. This allows us to keep all listeners intact and emit the events ourselves in `respond`.